File: android/app/src/main/java/app/releaf/mobile/data/sync/SyncPayloads.kt
/*
 * SyncPayloads.kt
 *
 * v2 Drive wire formats. Each entity kind gets a payload type, a
 * `toV2Payload()` mapper from the local Entity, and a `.toEntity(...)`
 * inverse for the download path.
 *
 * Rules:
 *   - Keys are snake_case and match the SQL column names byte-for-byte
 *     so a human reading the JSON can grep the schema for field meaning.
 *   - JSON-typed columns (contacts, locations, todos, attachments,
 *     sketch_strokes, sub_pages) land on Drive as embedded JSON arrays
 *     or objects — NOT as quoted strings. The payload type uses
 *     `JsonElement` for these fields; the `to`/`from` mappers parse the
 *     local string column on write and re-serialize it on read.
 *   - Drive payloads NEVER carry `dirty` or `drive_file_id` — those are
 *     local bookkeeping. The only Drive-side identity is the file's
 *     entity id + its checksum in the manifest.
 *   - `conflict_stub` is local-only too. It's a device's unresolved-
 *     merge record; peer devices don't know or care.
 *
 * Round-trip contract: for a row with `dirty = 0`, decoding its Drive
 * payload → re-encoding it must produce byte-identical canonical JSON
 * to what was uploaded. The test suite covers this for every kind.
 */

package app.releaf.mobile.data.sync

import app.releaf.mobile.data.notebook.BookSeriesEntity
import app.releaf.mobile.data.notebook.ChapterEntity
import app.releaf.mobile.data.notebook.NotebookEntity
import app.releaf.mobile.data.notebook.PageEntity
import app.releaf.mobile.data.notepad.NotepadEntry
import app.releaf.mobile.data.shelf.ShelfEntity
import app.releaf.mobile.data.task.TaskEntity
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement

/**
 * Default JSON configuration for sync. `ignoreUnknownKeys` lets a
 * newer writer's extra fields flow past an older reader without
 * blowing up (forward-compat for `minor` bumps per OPEN_QUESTIONS §5).
 */
val SyncJson: Json = Json {
    prettyPrint       = false
    ignoreUnknownKeys = true
    encodeDefaults    = true
}

// =====================================================================
// Notebook
// =====================================================================

@Serializable
data class NotebookPayloadV2(
    @SerialName("id")            val id: String,
    @SerialName("title")         val title: String,
    @SerialName("description")   val description: String? = null,
    @SerialName("color_hex")     val colorHex: String? = null,
    @SerialName("position")      val position: Long,
    @SerialName("archived_at")   val archivedAt: String? = null,
    /**
     * Shelf this book belongs to. Optional on the wire so old
     * clients that don't yet send it still round-trip; readers
     * default absent rows onto the "shelf-general" shelf.
     */
    @SerialName("shelf_id")      val shelfId: String? = null,
    @SerialName("series_id")     val seriesId: String? = null,
    @SerialName("volume_number") val volumeNumber: Int? = null,
    @SerialName("volume_name")   val volumeName: String? = null,
    @SerialName("created_at")    val createdAt: String,
    @SerialName("updated_at")    val updatedAt: String,
)

fun NotebookEntity.toV2Payload(): NotebookPayloadV2 = NotebookPayloadV2(
    id           = id,
    title        = title,
    description  = description,
    colorHex     = colorHex,
    position     = position,
    archivedAt   = archivedAt,
    shelfId      = shelfId,
    seriesId     = seriesId,
    volumeNumber = volumeNumber,
    volumeName   = this.volumeName,
    createdAt    = createdAt,
    updatedAt    = updatedAt,
)

fun NotebookPayloadV2.toEntity(driveFileId: String?): NotebookEntity = NotebookEntity(
    id           = id,
    title        = title,
    description  = description,
    colorHex     = colorHex,
    position     = position,
    shelfId      = shelfId ?: "shelf-general",
    seriesId     = seriesId,
    volumeNumber = volumeNumber ?: 1,
    volumeName   = this.volumeName,
    driveFileId  = driveFileId,
    createdAt    = createdAt,
    updatedAt    = updatedAt,
    dirty        = false,
    archivedAt   = archivedAt,
    deletedAt    = null,
)

// =====================================================================
// Shelf
// =====================================================================

@Serializable
data class ShelfPayloadV2(
    @SerialName("id")         val id: String,
    @SerialName("name")       val name: String,
    @SerialName("color_hex")  val colorHex: String? = null,
    @SerialName("position")   val position: Long,
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String,
)

fun ShelfEntity.toV2Payload(): ShelfPayloadV2 = ShelfPayloadV2(
    id        = id,
    name      = name,
    colorHex  = colorHex,
    position  = position,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun ShelfPayloadV2.toEntity(): ShelfEntity = ShelfEntity(
    id        = id,
    name      = name,
    colorHex  = colorHex,
    position  = position,
    createdAt = createdAt,
    updatedAt = updatedAt,
    dirty     = false,
    deletedAt = null,
)

// =====================================================================
// Book series
// =====================================================================

@Serializable
data class BookSeriesPayloadV2(
    @SerialName("id")         val id: String,
    @SerialName("shelf_id")   val shelfId: String,
    @SerialName("name")       val name: String,
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String,
)

fun BookSeriesEntity.toV2Payload(): BookSeriesPayloadV2 = BookSeriesPayloadV2(
    id        = id,
    shelfId   = shelfId,
    name      = name,
    createdAt = createdAt,
    updatedAt = updatedAt,
)

fun BookSeriesPayloadV2.toEntity(): BookSeriesEntity = BookSeriesEntity(
    id        = id,
    shelfId   = shelfId,
    name      = name,
    createdAt = createdAt,
    updatedAt = updatedAt,
    dirty     = false,
    deletedAt = null,
)

// =====================================================================
// Chapter
// =====================================================================

@Serializable
data class ChapterPayloadV2(
    @SerialName("id")          val id: String,
    @SerialName("notebook_id") val notebookId: String,
    @SerialName("title")       val title: String,
    @SerialName("description") val description: String? = null,
    @SerialName("position")    val position: Long,
    @SerialName("created_at")  val createdAt: String,
    @SerialName("updated_at")  val updatedAt: String,
)

fun ChapterEntity.toV2Payload(): ChapterPayloadV2 = ChapterPayloadV2(
    id          = id,
    notebookId  = notebookId,
    title       = title,
    description = description,
    position    = position,
    createdAt   = createdAt,
    updatedAt   = updatedAt,
)

fun ChapterPayloadV2.toEntity(driveFileId: String?): ChapterEntity = ChapterEntity(
    id          = id,
    notebookId  = notebookId,
    title       = title,
    description = description,
    position    = position,
    driveFileId = driveFileId,
    createdAt   = createdAt,
    updatedAt   = updatedAt,
    dirty       = false,
    deletedAt   = null,
)

// =====================================================================
// Page
// =====================================================================

@Serializable
data class PagePayloadV2(
    @SerialName("id")             val id: String,
    @SerialName("chapter_id")     val chapterId: String,
    @SerialName("project_id")     val projectId: String? = null,
    @SerialName("template_id")    val templateId: String? = null,
    @SerialName("title")          val title: String? = null,
    @SerialName("notes")          val notes: String,
    @SerialName("contacts")       val contacts: JsonElement,
    @SerialName("locations")      val locations: JsonElement,
    @SerialName("todos")          val todos: JsonElement,
    @SerialName("attachments")    val attachments: JsonElement,
    @SerialName("sketch_strokes") val sketchStrokes: JsonElement,
    @SerialName("sub_pages")      val subPages: JsonElement,
    @SerialName("position")       val position: Long,
    @SerialName("created_at")     val createdAt: String,
    @SerialName("updated_at")     val updatedAt: String,
)

fun PageEntity.toV2Payload(): PagePayloadV2 = PagePayloadV2(
    id            = id,
    chapterId     = chapterId,
    projectId     = projectId,
    templateId    = templateId,
    title         = title,
    notes         = notes,
    contacts      = parseJsonArrayOrEmpty(contacts),
    locations     = parseJsonArrayOrEmpty(locations),
    todos         = parseJsonArrayOrEmpty(todos),
    attachments   = parseJsonArrayOrEmpty(attachments),
    sketchStrokes = parseJsonArrayOrEmpty(sketchStrokes),
    subPages      = parseJsonArrayOrEmpty(subPages),
    position      = position,
    createdAt     = createdAt,
    updatedAt     = updatedAt,
)

fun PagePayloadV2.toEntity(driveFileId: String?): PageEntity = PageEntity(
    id            = id,
    chapterId     = chapterId,
    projectId     = projectId,
    templateId    = templateId,
    title         = title,
    notes         = notes,
    contacts      = SyncJson.encodeToString(JsonElement.serializer(), contacts),
    locations     = SyncJson.encodeToString(JsonElement.serializer(), locations),
    todos         = SyncJson.encodeToString(JsonElement.serializer(), todos),
    attachments   = SyncJson.encodeToString(JsonElement.serializer(), attachments),
    sketchStrokes = SyncJson.encodeToString(JsonElement.serializer(), sketchStrokes),
    subPages      = SyncJson.encodeToString(JsonElement.serializer(), subPages),
    position      = position,
    conflictStub  = null,
    driveFileId   = driveFileId,
    createdAt     = createdAt,
    updatedAt     = updatedAt,
    dirty         = false,
    deletedAt     = null,
)

// =====================================================================
// NotepadEntry
// =====================================================================

@Serializable
data class NotepadEntryPayloadV2(
    @SerialName("id")                  val id: String,
    @SerialName("user_id")             val userId: String,
    @SerialName("entry_date")          val entryDate: String,
    @SerialName("project_id")          val projectId: String? = null,
    @SerialName("title")               val title: String? = null,
    @SerialName("notes")               val notes: String,
    @SerialName("contacts")            val contacts: JsonElement,
    @SerialName("locations")           val locations: JsonElement,
    @SerialName("todos")               val todos: JsonElement,
    @SerialName("attachments")         val attachments: JsonElement,
    @SerialName("sketch_strokes")      val sketchStrokes: JsonElement,
    @SerialName("sub_pages")           val subPages: JsonElement,
    @SerialName("allow_blank_content") val allowBlankContent: Boolean,
    @SerialName("created_at")          val createdAt: String,
    @SerialName("updated_at")          val updatedAt: String,
)

fun NotepadEntry.toV2Payload(): NotepadEntryPayloadV2 = NotepadEntryPayloadV2(
    id                = id,
    userId            = userId,
    entryDate         = entryDate,
    projectId         = projectId,
    title             = title,
    notes             = notes,
    contacts          = parseJsonArrayOrEmpty(contacts),
    locations         = parseJsonArrayOrEmpty(locations),
    todos             = parseJsonArrayOrEmpty(todos),
    attachments       = parseJsonArrayOrEmpty(attachments),
    sketchStrokes     = parseJsonArrayOrEmpty(sketchStrokes),
    subPages          = parseJsonArrayOrEmpty(subPages),
    allowBlankContent = allowBlankContent,
    createdAt         = createdAt,
    updatedAt         = updatedAt,
)

fun NotepadEntryPayloadV2.toEntity(driveFileId: String?): NotepadEntry = NotepadEntry(
    id                = id,
    userId            = userId,
    entryDate         = entryDate,
    projectId         = projectId,
    title             = title,
    notes             = notes,
    contacts          = SyncJson.encodeToString(JsonElement.serializer(), contacts),
    locations         = SyncJson.encodeToString(JsonElement.serializer(), locations),
    todos             = SyncJson.encodeToString(JsonElement.serializer(), todos),
    attachments       = SyncJson.encodeToString(JsonElement.serializer(), attachments),
    sketchStrokes     = SyncJson.encodeToString(JsonElement.serializer(), sketchStrokes),
    subPages          = SyncJson.encodeToString(JsonElement.serializer(), subPages),
    allowBlankContent = allowBlankContent,
    conflictStub      = null,
    driveFileId       = driveFileId,
    createdAt         = createdAt,
    updatedAt         = updatedAt,
    dirty             = false,
    deletedAt         = null,
)

// =====================================================================
// Task
// =====================================================================

@Serializable
data class TaskPayloadV2(
    @SerialName("id")           val id: String,
    @SerialName("user_id")      val userId: String,
    @SerialName("title")        val title: String,
    @SerialName("description")  val description: String? = null,
    @SerialName("due_date")     val dueDate: String? = null,
    @SerialName("completed")    val completed: Boolean,
    @SerialName("completed_at") val completedAt: String? = null,
    @SerialName("priority")     val priority: Int,
    @SerialName("status")       val status: String,
    @SerialName("created_at")   val createdAt: String,
    @SerialName("updated_at")   val updatedAt: String,
)

fun TaskEntity.toV2Payload(): TaskPayloadV2 = TaskPayloadV2(
    id          = id,
    userId      = userId,
    title       = title,
    description = description,
    dueDate     = dueDate,
    completed   = completed,
    completedAt = completedAt,
    priority    = priority,
    status      = status,
    createdAt   = createdAt,
    updatedAt   = updatedAt,
)

fun TaskPayloadV2.toEntity(): TaskEntity = TaskEntity(
    id          = id,
    userId      = userId,
    title       = title,
    description = description,
    dueDate     = dueDate,
    completed   = completed,
    completedAt = completedAt,
    priority    = priority,
    status      = status,
    createdAt   = createdAt,
    updatedAt   = updatedAt,
    dirty       = false,
    deletedAt   = null,
)

// =====================================================================
// Helpers
// =====================================================================

/**
 * Parse a locally-stored JSON string column into a `JsonElement`. On
 * parse failure, return an empty array — the row is still uploadable
 * and the bad column is surfaced as `[]` on Drive, which is safer than
 * aborting the whole sync pass. Emits to logs via `println` on failure;
 * replace with a real logger when we get one.
 */
private fun parseJsonArrayOrEmpty(raw: String?): JsonElement {
    if (raw.isNullOrEmpty()) return JsonArray(emptyList())
    return try {
        SyncJson.parseToJsonElement(raw)
    } catch (_: Exception) {
        JsonArray(emptyList())
    }
}

File: ios/Releaf/Data/Sync/SyncPayloads.swift
/*
 * SyncPayloads.swift
 *
 * v2 Drive wire formats for iOS. Mirror of Android's `SyncPayloads.kt`
 * — the two emit byte-identical canonical JSON for identical row data.
 *
 * Shape rules:
 *   - snake_case keys matching SQL columns.
 *   - JSON-typed columns (contacts, locations, todos, attachments,
 *     sketch_strokes, sub_pages) land as embedded JSON values, not
 *     quoted strings. We use `JSONAny` (a thin wrapper over Any) to
 *     pass these through Codable.
 *   - Drive payloads never carry `dirty` or `drive_file_id`.
 *   - `conflict_stub` is local-only.
 */

import Foundation

// =====================================================================
// JSONAny — pass-through JSON value for Codable
// =====================================================================

/// Pass-through Codable wrapper for an embedded JSON value (array,
/// object, string, number, bool, null). Used for payload fields that
/// carry the on-disk JSON string verbatim.
public struct JSONAny: Codable, Equatable, Sendable {
    public let value: Any

    public init(_ value: Any) { self.value = value }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self.value = NSNull()
        } else if let b = try? container.decode(Bool.self) {
            self.value = b
        } else if let i = try? container.decode(Int64.self) {
            self.value = i
        } else if let d = try? container.decode(Double.self) {
            self.value = d
        } else if let s = try? container.decode(String.self) {
            self.value = s
        } else if let a = try? container.decode([JSONAny].self) {
            self.value = a.map(\.value)
        } else if let o = try? container.decode([String: JSONAny].self) {
            var m = [String: Any]()
            for (k, v) in o { m[k] = v.value }
            self.value = m
        } else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "JSONAny: unsupported value"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try encodeAny(value, into: &container)
    }

    public static func == (lhs: JSONAny, rhs: JSONAny) -> Bool {
        // Round-trip compare via canonical JSON bytes.
        CanonicalJson.encodeToString(lhs.value) ==
            CanonicalJson.encodeToString(rhs.value)
    }

    /// Parse a raw JSON string into a JSONAny. On parse failure,
    /// returns an empty array — mirrors Android's behavior so the bad
    /// row uploads as `[]` rather than blocking the whole sync pass.
    public static func parseOrEmptyArray(_ raw: String?) -> JSONAny {
        guard let raw = raw, !raw.isEmpty,
              let data = raw.data(using: .utf8),
              let tree = try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
        else {
            return JSONAny([Any]())
        }
        return JSONAny(tree)
    }

    /// Re-serialize this tree into a compact JSON string for local
    /// column storage. Uses `sortedKeys` so repeated round-trips are
    /// idempotent.
    public func toCompactString() -> String {
        guard let data = try? JSONSerialization.data(
            withJSONObject: value,
            options: [.sortedKeys, .fragmentsAllowed]
        ) else {
            return "[]"
        }
        return String(data: data, encoding: .utf8) ?? "[]"
    }
}

/// Recursive encoder for a Foundation JSON tree.
private func encodeAny(_ any: Any, into container: inout SingleValueEncodingContainer) throws {
    if any is NSNull { try container.encodeNil(); return }

    switch any {
    case let b as Bool:     try container.encode(b)
    case let i as Int:      try container.encode(Int64(i))
    case let i as Int64:    try container.encode(i)
    case let d as Double:   try container.encode(d)
    case let s as String:   try container.encode(s)
    case let arr as [Any]:
        try container.encode(arr.map { JSONAny($0) })
    case let nsarr as NSArray:
        try container.encode(Array(nsarr).map { JSONAny($0) })
    case let dict as [String: Any]:
        try container.encode(dict.mapValues { JSONAny($0) })
    case let nsdict as NSDictionary:
        var m = [String: JSONAny]()
        for case let (k as String, v) in nsdict { m[k] = JSONAny(v) }
        try container.encode(m)
    case let n as NSNumber:
        let type = String(cString: n.objCType)
        if type == "c" || type == "B" {
            try container.encode(n.boolValue)
        } else {
            let d = n.doubleValue
            if d.truncatingRemainder(dividingBy: 1) == 0,
               d >= Double(Int64.min), d <= Double(Int64.max) {
                try container.encode(Int64(d))
            } else {
                try container.encode(d)
            }
        }
    default:
        try container.encodeNil()
    }
}

// =====================================================================
// Notebook
// =====================================================================

public struct NotebookPayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let title: String
    public let description: String?
    public let colorHex: String?
    public let position: Int64
    public let archivedAt: String?
    /// Shelf this book belongs to. Optional on the wire so old
    /// clients that don't yet send it can still round-trip; readers
    /// default absent rows onto the "shelf-general" shelf.
    public let shelfId: String?
    public let seriesId: String?
    public let volumeNumber: Int?
    public let volumeName: String?
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id, title, description
        case colorHex     = "color_hex"
        case position
        case archivedAt   = "archived_at"
        case shelfId      = "shelf_id"
        case seriesId     = "series_id"
        case volumeNumber = "volume_number"
        case volumeName   = "volume_name"
        case createdAt    = "created_at"
        case updatedAt    = "updated_at"
    }
}

public extension NotebookEntity {
    func toV2Payload() -> NotebookPayloadV2 {
        NotebookPayloadV2(
            id: id,
            title: title,
            description: nil, // iOS NotebookEntity has no description column yet
            colorHex: colorHex,
            position: position,
            archivedAt: nil,
            shelfId: shelfId,
            seriesId: seriesId,
            volumeNumber: volumeNumber,
            volumeName: volumeName,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension NotebookPayloadV2 {
    func toEntity(driveFileId: String?) -> NotebookEntity {
        NotebookEntity(
            id: id,
            title: title,
            colorHex: colorHex,
            position: position,
            shelfId: shelfId ?? "shelf-general",
            seriesId: seriesId,
            volumeNumber: volumeNumber ?? 1,
            volumeName: volumeName,
            driveFileId: driveFileId,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}

// =====================================================================
// Shelf
// =====================================================================

public struct ShelfPayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let name: String
    public let colorHex: String?
    public let position: Int64
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id, name
        case colorHex  = "color_hex"
        case position
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

public extension ShelfEntity {
    func toV2Payload() -> ShelfPayloadV2 {
        ShelfPayloadV2(
            id: id,
            name: name,
            colorHex: colorHex,
            position: position,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension ShelfPayloadV2 {
    func toEntity() -> ShelfEntity {
        ShelfEntity(
            id: id,
            name: name,
            colorHex: colorHex,
            position: position,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}

// =====================================================================
// Book series
// =====================================================================

public struct BookSeriesPayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let shelfId: String
    public let name: String
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id
        case shelfId   = "shelf_id"
        case name
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

public extension BookSeriesEntity {
    func toV2Payload() -> BookSeriesPayloadV2 {
        BookSeriesPayloadV2(
            id: id,
            shelfId: shelfId,
            name: name,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension BookSeriesPayloadV2 {
    func toEntity() -> BookSeriesEntity {
        BookSeriesEntity(
            id: id,
            shelfId: shelfId,
            name: name,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}

// =====================================================================
// Chapter
// =====================================================================

public struct ChapterPayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let notebookId: String
    public let title: String
    public let description: String?
    public let position: Int64
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id
        case notebookId = "notebook_id"
        case title, description
        case position
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

public extension ChapterEntity {
    func toV2Payload() -> ChapterPayloadV2 {
        ChapterPayloadV2(
            id: id,
            notebookId: notebookId,
            title: title,
            description: nil, // iOS ChapterEntity has no description column yet
            position: position,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension ChapterPayloadV2 {
    func toEntity(driveFileId: String?) -> ChapterEntity {
        ChapterEntity(
            id: id,
            notebookId: notebookId,
            title: title,
            position: position,
            driveFileId: driveFileId,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}

// =====================================================================
// Page
// =====================================================================

public struct PagePayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let chapterId: String
    public let projectId: String?
    public let templateId: String?
    public let title: String?
    public let notes: String
    public let contacts: JSONAny
    public let locations: JSONAny
    public let todos: JSONAny
    public let attachments: JSONAny
    public let sketchStrokes: JSONAny
    public let subPages: JSONAny
    public let position: Int64
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id
        case chapterId     = "chapter_id"
        case projectId     = "project_id"
        case templateId    = "template_id"
        case title, notes, contacts, locations, todos, attachments
        case sketchStrokes = "sketch_strokes"
        case subPages      = "sub_pages"
        case position
        case createdAt     = "created_at"
        case updatedAt     = "updated_at"
    }
}

public extension PageEntity {
    func toV2Payload() -> PagePayloadV2 {
        PagePayloadV2(
            id: id,
            chapterId: chapterId,
            projectId: projectId,
            templateId: templateId,
            title: title,
            notes: notes,
            contacts: JSONAny.parseOrEmptyArray(contacts),
            locations: JSONAny.parseOrEmptyArray(locations),
            todos: JSONAny.parseOrEmptyArray(todos),
            attachments: JSONAny.parseOrEmptyArray(attachments),
            sketchStrokes: JSONAny([Any]()),
            subPages: JSONAny([Any]()),
            position: position,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension PagePayloadV2 {
    func toEntity(driveFileId: String?) -> PageEntity {
        PageEntity(
            id: id,
            chapterId: chapterId,
            projectId: projectId,
            templateId: templateId,
            title: title,
            notes: notes,
            contacts: contacts.toCompactString(),
            locations: locations.toCompactString(),
            todos: todos.toCompactString(),
            attachments: attachments.toCompactString(),
            position: position,
            conflictStub: nil,
            driveFileId: driveFileId,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}

// =====================================================================
// NotepadEntry
// =====================================================================

public struct NotepadEntryPayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let userId: String
    public let entryDate: String
    public let projectId: String?
    public let title: String?
    public let notes: String
    public let contacts: JSONAny
    public let locations: JSONAny
    public let todos: JSONAny
    public let attachments: JSONAny
    public let sketchStrokes: JSONAny
    public let subPages: JSONAny
    public let allowBlankContent: Bool
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id
        case userId            = "user_id"
        case entryDate         = "entry_date"
        case projectId         = "project_id"
        case title, notes, contacts, locations, todos, attachments
        case sketchStrokes     = "sketch_strokes"
        case subPages          = "sub_pages"
        case allowBlankContent = "allow_blank_content"
        case createdAt         = "created_at"
        case updatedAt         = "updated_at"
    }
}

public extension NotepadEntry {
    func toV2Payload() -> NotepadEntryPayloadV2 {
        NotepadEntryPayloadV2(
            id: id,
            userId: userId,
            entryDate: entryDate,
            projectId: projectId,
            title: title,
            notes: notes,
            contacts: JSONAny.parseOrEmptyArray(contacts),
            locations: JSONAny.parseOrEmptyArray(locations),
            todos: JSONAny.parseOrEmptyArray(todos),
            attachments: JSONAny.parseOrEmptyArray(attachments),
            sketchStrokes: JSONAny([Any]()),
            subPages: JSONAny([Any]()),
            allowBlankContent: allowBlankContent,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension NotepadEntryPayloadV2 {
    func toEntity(driveFileId: String?) -> NotepadEntry {
        NotepadEntry(
            id: id,
            userId: userId,
            entryDate: entryDate,
            projectId: projectId,
            title: title,
            notes: notes,
            contacts: contacts.toCompactString(),
            locations: locations.toCompactString(),
            todos: todos.toCompactString(),
            attachments: attachments.toCompactString(),
            allowBlankContent: allowBlankContent,
            conflictStub: nil,
            driveFileId: driveFileId,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}

// =====================================================================
// Task
// =====================================================================

public struct TaskPayloadV2: Codable, Equatable, Sendable {
    public let id: String
    public let userId: String
    public let title: String
    public let description: String?
    public let dueDate: String?
    public let completed: Bool
    public let completedAt: String?
    public let priority: Int
    public let createdAt: String
    public let updatedAt: String

    public enum CodingKeys: String, CodingKey {
        case id
        case userId      = "user_id"
        case title, description
        case dueDate     = "due_date"
        case completed
        case completedAt = "completed_at"
        case priority
        case createdAt   = "created_at"
        case updatedAt   = "updated_at"
    }
}

public extension TaskRecord {
    func toV2Payload() -> TaskPayloadV2 {
        TaskPayloadV2(
            id: id,
            userId: userId,
            title: title,
            description: description,
            dueDate: dueDate,
            completed: completed,
            completedAt: completedAt,
            priority: priority,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

public extension TaskPayloadV2 {
    func toEntity() -> TaskRecord {
        TaskRecord(
            id: id,
            userId: userId,
            title: title,
            description: description,
            dueDate: dueDate,
            completed: completed,
            completedAt: completedAt,
            priority: priority,
            createdAt: createdAt,
            updatedAt: updatedAt,
            dirty: false,
            deletedAt: nil
        )
    }
}
